source/Octopus.Cli/Commands/Deployment/TaskOutputProgressPrinter.cs

```diff
@@ -144,5 +154,10 @@ static string ConvertToTeamCityMessageStatus(string category)
             }
             return "NORMAL";
         }
+
+        static bool IsTeamCityServiceMessage(string line)
```

Comment:
This feels a bit too teamcity specific... the code above has abstracted away the service messages (well, mostly... except for the `ConvertToTeamCityMessageStatus`), but this brings it back in.

Can it/should it be done in `LogExtensions`?

Will we face the same issue with TFS service messages?

Reply:
> This feels a bit too teamcity specific... the code above has abstracted away the service messages (well, mostly... except for the ConvertToTeamCityMessageStatus), but this brings it back in.

That's because _it is_ TeamCity specific, for TFS/VSTS/ADOps we just write the messages as `Information`.

> Can it/should it be done in LogExtensions?

We could, but we'd have to inspect the `new { text = line, status = ConvertToTeamCityMessageStatus(logEntry.Category) }` object passed in to figure out if the line is a teamcity service message or not, instead of just checking the `line` we're currently processing. 

And since we already have the `ConvertToTeamCityMessageStatus` (and also TeamCity/VSTS specific *render* methods) I don't see what benefits it would be to move it into `LogExtensions` with the need to inspect the passed in object to figure out what to do...

> Will we face the same issue with TFS service messages?

No, see point made in first answer

Reply:
What about https://github.com/OctopusDeploy/OctopusCLI/blob/master/source/Octopus.Cli/Diagnostics/LogExtensions.cs#L134? that's writing a specific `##vso` tag.

~I'm commenting more on the "where" rather than the "what" - it feels the abstraction levels are wrong. `TaskOutputProgressPrinter.cs` is (mostly) build server agnostic - that logic is (mostly) encapsulated in LogExtensions. This seems to spread the knowledge about build servers around the code.~

~What if it did:~

```
if (LogExtensions.IsServiceMessage(line))
{
    commandOutputProvider.Information(line);
}
else
{
    commandOutputProvider.ServiceMessage("message", new { text = line, status = ConvertToTeamCityMessageStatus(logEntry.Category) });
}
```

and in LogExtensions:
```
static bool IsServiceMessage(string line)
{
    return line.StartsWith("##teamcity", StringComparison.OrdinalIgnoreCase) || 
           line.StartsWith("##vso", StringComparison.OrdinalIgnoreCase) || ; 
}
```

(the `##vso` line may not be necessary, but then will only have to modify the LogExtensions class if anything changes around service messages)

Hmmm. And then finally re-reading your comment
> And since we already have the ConvertToTeamCityMessageStatus (and also TeamCity/VSTS specific render methods) 
I've looked at the rest of the class, and I think i take it all back... this class is pretty tightly coupled to build servers. That's probably a candidate for a refactor at some point.

:nothingtoseehere:

Reply:
These methods seem to pretty clear separate TC and VSO, and given this one has TeamCity clearly in the name I don't know that it needs to be checking for vso things. 

```
var isTeamCityServiceMessage = line.StartsWith("##teamcity", StringComparison.OrdinalIgnoreCase);
if (isTeamCityServiceMessage)
...
```
may suffice here?

Is the concern here that the vso method may need to do something similar to make sure it passes what are meant to be service messages from the script through correctly?